File: cpp/structures/Stack.cpp

#include <iostream>
class Node {
 public:
  int value;
  Node* next;

  Node(int value) {
    this->value = value;
    next = nullptr;
  }
};

class Stack {
 private:
  Node* top;
  int height;

 public:
  Stack(int value) {
    Node* newNode = new Node(value);
    top = newNode;
    height = 1;
  }

  void push(int value) {
    Node* newNode = new Node(value);

    newNode->next = top;
    top = newNode;

    height++;
  }

  Node* pop() {
    if (height == 0) return nullptr;

    Node* temp = top;
    top = top->next;
    temp->next = nullptr;

    height--;
    return temp;
  }

  void print() {
    Node* temp = top;
    while (temp) {
      std::cout << temp->value << std::endl;
      temp = temp->next;
    }
  }

  void getTop() { std::cout << "Top: " << top->value << std::endl; }
  void getHeight() { std::cout << "Height: " << height << std::endl; }
};

int main(int argc, char* argv[]) {
  Stack* stack = new Stack(1);
  stack->push(2);
  stack->push(3);

  stack->print();

  return 0;
}

// clang++ -std=c++20 Stack.cpp -o Stack ; ./Stack

File: python/structures/stack.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


class Stack:
    def __init__(self, value):
        node = Node(value)
        self.top = node
        self.height = 1

    def push(self, value):
        node = Node(value)
        if self.height == 0:
            self.top = node
        else:
            node.next = self.top
            self.top = node

        self.height += 1

    def pop(self) -> Node | None:
        if self.height == 0:
            return None

        temp = self.top

        self.top = self.top.next
        temp.next = None

        self.height -= 1
        return temp

    def print(self):
        temp = self.top
        while temp is not None:
            print(temp.value)
            temp = temp.next


if __name__ == "__main__":
    stack = Stack(2)
    stack.push(1)

    stack.print()
